// v/templates/new/js/models/schedule.js
define(function(require, exports, module) {
    'use strict'

    var API = M.config.API,
        scheduleURL = 'poi/movieshows.json?',
        DetailModel = require('/js/models/detail'),
        _ = require('_'),
        Backbone = require('Backbone')

    var ScheduleModel = Backbone.Model.extend({
        urlRoot: function() {
            // poi/movieshows.json?pois=1466992&dt=2012-10-09
            return (API + scheduleURL + 'pois=' + initData.poiid + '&dt=' + (new Date()).toISOString().split('T')[0])
        },
        parse: function(res) {
            var ret = {
                temp: {
                    schedule: []
                }
            }

            res[initData.poiid].forEach(function(mov, index) {
                var movie = ret.temp.schedule[index] = {
                    name: mov.nm,
                    duration: mov.dur,
                    types: []
                }

                mov.plist.forEach(function(m) {
                    var type = (m.tp || '') + (m.lang || '')

                    if (movie.types.indexOf(type) === -1) {
                        movie.types.push(type)
                        movie[type] = []
                    }
                    movie[type].push({ time: m.tm })
                })
            })

            return ret
        },
    })

    module.exports = new ScheduleModel
})
