// v/templates/new/js/views/detail.js
define(function(require, exports, module) {
    'use strict'

    var DetailModel = require('/js/models/detail'),
        Handlebars = require('Handlebars'),
        Backbone = require('Backbone'),
        Tmpl = require('/js/template'),
        Map = require('http://ditu.google.cn/maps/api/js?v=3.9&sensor=false&language=zh-CN&callback=M.app.DetailView.initMap')

    module.exports = Backbone.View.extend({
        el: '#merchant',
        template: Handlebars.compile(Tmpl.detail),
        initialize: function() {
            this.$el.prepend(this.template(DetailModel.toJSON()))

            return this
        },
        initMap: function() {
            var latlng = new google.maps.LatLng(DetailModel.get("lat"), DetailModel.get("lng")),
                mapOptions = {
                    center: latlng,
                    disableDefaultUI: true,
                    mapTypeId: google.maps.MapTypeId.ROADMAP,
                    zoom: 15
                },
                map = new google.maps.Map(document.getElementById('map_canvas'), mapOptions),
                markerOptions = {
                    position: latlng,
                    map: map,
                    title: DetailModel.get('name'),
                    visible: true
                },
                marker = new google.maps.Marker( markerOptions )

            // create Home Control
            var homeControlDiv = document.createElement('div');
            var homeControl = new HomeControl(homeControlDiv, map, latlng, DetailModel.get('name'));

            homeControlDiv.index = 1;
            map.controls[google.maps.ControlPosition.TOP_LEFT].push(homeControlDiv);
            // 手动释放内存
            delete M.app.DetailView

            return {
                map: map,
                marker: marker
            }
        }
    })

    /**
     * The Home Control adds a control to the map that
     * simply to returns the user to target
     */
    function HomeControl(controlDiv, map, latlng, name) {

        controlDiv.style.padding = '5px';

        var controlUI = document.createElement('div');
        controlUI.style.cssText = "background:white; border: 2px solid #000; text-align:center;";
        controlDiv.appendChild(controlUI);

        var controlText = document.createElement('div');
        controlText.style.cssText = 'font-size:12px; padding: 0 4px; max-width:129px; white-space:nowrap; overflow:hidden; text-overflow:ellipsis;';
        controlText.innerHTML = '<b>' + name + '</b>';
        controlUI.appendChild(controlText);

        google.maps.event.addDomListener(controlUI, 'click', function() {
          map.setCenter(latlng)
        });

    }
})
