// v/templates/new/js/views/nearbydeals.js
define(function(require, exports, module) {
    'use strict'

    var NearbyDealsModel = require('/js/models/nearbydeals'),
        Backbone = require('Backbone'),
        _ = require('_'),
        Handlebars = require('Handlebars'),
        Tmpl = require('/js/template')

    module.exports = Backbone.View.extend({
        el: '#merchant',
        template: Handlebars.compile(Tmpl.nearby_deals),
        model: NearbyDealsModel,
        events: {
            'click .more-deals': 'getMoreDeals'
        },
        initialize: function() {
            _.bindAll(this, 'buildView')
            NearbyDealsModel.on('change', this.buildView, this)

            return this
        },
        buildView: function() {
            var $nearby = this.$el.find('#nearby_deals'),
                nearbyDeals = NearbyDealsModel.toJSON()

            // '更多'按钮被点击时
            if ($nearby.length !== 0) {
                $nearby.replaceWith(this.template(nearbyDeals));

            // 初始化‘附近团购’时
            } else {
                if (nearbyDeals.data.length > 3) {
                    nearbyDeals.data = nearbyDeals.data.splice(0, 3)
                    nearbyDeals.data.push({mname: '更多'})
                }
                if (!initData.hasBought) {
                    this.$el.find('#loading_tips').remove()
                    this.$el.append(this.template(nearbyDeals))
                } else {
                    this.$el.find('#loading_tips').before(this.template(nearbyDeals))
                }
            }
        },
        getMoreDeals: function() {
            var $arrow = this.$el.find('.more-deals .icon-right')

            if (!this.hasTapped) {
                this.hasTapped = true;
                $arrow.addClass('loading')

                NearbyDealsModel.fetch({
                    dataType: "jsonp",
                    url: NearbyDealsModel.urlRoot() + "&limit=15",
                    success: function() {
                        $arrow.removeClass('loading')
                    },
                    error: function(xhr, errorType, error) {
                        this.hasTapped = false
                        $arrow.removeClass('loading')
                        M.log({
                            msg: "**EVE**fetch more nearby deals data error | " + JSON.stringify(error),
                            url: merchant.NearbyDealsModel.urlRoot()
                        })
                    }
                })
            }
        }
    })
})
